### ISSUE TITLE
BDPT: importance should have pixel norm

### ISSUE THREAD

**Issue by rainbow-app** [mmp/pbrt-v4#456]
Hello!

**UPD: 2024 nov 12**. Three posts here:
* this post provides an example where pixel norm is clearly advantageous
* 2nd post admits the opposite
* 3d post is the discussion

---

Writing to v4, but using v3, I'm sure v4 is affected.

Importance for a single pixel is the same as in textbook, except the area in denominator is pixel area. Brief derivation is here: https://github.com/mmp/pbrt-v3/issues/336#issuecomment-2391946322. (I tried to explain it to someone, thinking I understand it...). It's easy to understand: it is pixels that measure, not the sensor. There's no measurement the whole sensor provides, so no importance for the sensor.

A test case, where the pixel norm is clearly better:
![wnorm-diagram](https://github.com/user-attachments/assets/744deb32-4f78-4144-8799-e5ad5a37e88d)

To make pixel norm:
* PerspectiveCamera::PerspectiveCamera
A=.../(res.x\*res.y);
* make splats per-pixel, in Film::WriteImage:
int resN=croppedPixelBounds.Area();
...
rgb[3 * offset] += .../resN; // and two other lines

Result, coneangle=30, spp=1, sensor-norm vs. pixel-norm:
![wnorm-ws](https://github.com/user-attachments/assets/0c8c8a6a-5954-4434-806d-10ba24f37d9d)
![wnorm-wp](https://github.com/user-attachments/assets/84640cdb-c774-426e-9531-c94b09922bae)

-----
How to reproduce

Some hard-coded numbers, for resolution 11x11:
size of area that maps to 1 pixel = pxsz = tan(60\*pi/180/2)/5.5 = 0.1050 (meters)
x coord where we point the light at: xpos = pxsz\*4 = 0.4199
x coord where the light is: pxsz\*4\*2 = 0.8398
distance to the illiminated point: dist = sqrt(xpos^2+1) = 1.0846
coneangle to cover that area (1 pixel): coneangle = pxsz/2\*(1/dist)/dist\*180/pi = 2.5565

Sampler "random" "integer pixelsamples" [1]
Integrator "bdpt" "integer maxdepth" [1]

LookAt 0 0 -1   0 0 0   0 1 0
Camera "perspective" "float fov" [60]
Film "image" "integer xresolution" [256] "integer yresolution" [256]
    "string filename" "wnorm.pfm"

WorldBegin

Material "matte" "color Kd" [.5 .5 .5] "float sigma" [0]

Shape "trianglemesh" "point P" [-1 -1 0  -1 1 0  1 1 0  1 -1 0]
        "integer indices" [ 0 1 2 2 3 0]

LightSource "spot" "point from" [0.8398 0 -1] "point to" [0.4199 0 0]
  "float coneangle" [30] "color I" [5 5 5]

WorldEnd

**Comment by rainbow-app** [comment]
I can't update images in old post, so creating a new one.

My original fix was incomplete, here's the correct fixed image:
![wnorm2](https://github.com/user-attachments/assets/c996df4a-af8a-4c92-8d03-a50aff1a441c)

---

The new norm significantly affects images for maxDepth>1. Here are the images of the scene from my old [cropwindow](https://github.com/mmp/pbrt-v4/issues/347) issue (downsized). Vanilla and fixed.
![z-ws-r](https://github.com/user-attachments/assets/f00ce06b-d186-4d8f-b38d-519fb72703b7)
![z-wp2-r](https://github.com/user-attachments/assets/9730d19f-ee92-49d1-a1e9-bc1c0c9dd3cb)
(need to think about this difference...)

**Comment by rainbow-app** [comment]
Ok, I think I finally figured it. (Long time ago I immediately noticed that textbook is very vague on this importance topic, but I thought I understood it, I didn't.)

To summarize (the way I see it):

There's no single = "overall" = whole-sensor importance. We have many importances for each pixel. Pixel's importance should have pixel area in the denominator.

The t=1 strategy doesn't (usually) give sample(s) for the given pixel. Hence, a new theory is required that considers the whole sensor, or an ad-hoc rule on how to add  the light image (=splats). Veach describes (vaguely, I'd say) in 10.3.4.3 (p.313) how he summed light and camera images, but that's an ad-hoc approach, not a proper solution.

Consider example case: fixed pixel area, some bounded scene in the FOV, and resolution goes to infinity (fov->180*). The number of splats -> infinity, so light image (the scene at the center) becomes statistically superb, w/o noise. However, in Veach theory each pixel is separate=single, so his MIS weights will not account for the superb noise-free light image. So we need to correct MIS weights (with W per pixel). How?

Basic argument (that I came up with, also vague): for the ad-hoc rule, we are not interested in separate pixel measuremenst, we sum the *whole* light image; splats are everywhere, but we are monitoring their contribution only in the sensor area. So a reasonable attempt would be to change the camera pdf density to be wrt sensor area -- for MIS purposes only. It turnes out (from some formulas) that this ad-hoc change does the above job pretty well. Notice that we don't need per-sensor "overall" importance, just camera pdf.

In pbrt, camera pdf is always per whole-sensor, so you have the MIS weights for summing of light and camera images (erroneously) correct. The *measurements* on the other hand are always in the form W/p, and so the erroneous per-sensor area cancels, and result is as if you'd use correct area (per pixel). And in one place the ratio W/p is further (analytically) computed into beta=1.

I implemented "my way" (it's merely a re-shuffling of terms in your formulas), works the same.

I'm leaving this open to attract attention, but feel free to close it.
